ocaml/xenopsd/c_stubs/xenctrlext_stubs.c

```diff
@@ -394,6 +409,135 @@ CAMLprim value stub_xenctrlext_cputopoinfo(value xch)
 	CAMLreturn(result);
 }
 
+CAMLprim value stub_xenforeignmemory_open(value logger, value open_flags)
+{
+        CAMLparam2(logger, open_flags);
+        struct xentoollog_logger *log_handle = NULL;
+        struct xenforeignmemory_handle *fmem;
+        CAMLlocal1(result);
+
+        if(Is_some(logger)) {
+                log_handle = Xtl_val(Some_val(logger));
+        }
+
+        fmem = xenforeignmemory_open(log_handle, Int_val(open_flags));
+
+        if(fmem == NULL) {
+                caml_failwith("Error when opening foreign memory handle");
+        }
+
+        result = caml_alloc(1, Abstract_tag);
+        Xfm_val(result) = fmem;
+
+        CAMLreturn(result);
+}
+
+CAMLprim value stub_xenforeignmemory_close(value fmem)
+{
+        CAMLparam1(fmem);
+        int retval;
+
+        if(Xfm_val(fmem) == NULL) {
+                caml_invalid_argument(
+                        "Error: cannot close NULL foreign memory handle");
+        }
+
+        retval = xenforeignmemory_close(Xfm_val(fmem));
+
+        if(retval < 0) {
+                caml_failwith("Error when closing foreign memory handle");
+        }
+
+        // Protect against double close
+        Xfm_val(fmem) = NULL;
+
+        CAMLreturn(Val_unit);
+}
+
+CAMLprim value stub_xenforeignmemory_map(value fmem, value dom,
+        value prot_flags, value pages)
+{
+        CAMLparam4(fmem, dom, prot_flags, pages);
+        CAMLlocal2(cell, result);
+        size_t i, pages_length;
+        xen_pfn_t *arr;
+        int prot, the_errno;
+        void *retval;
+
+        if (Field(prot_flags, 0) == Val_false &&
```

Comment:
I am wondering if the first `if` is needed. Could we start with `prot = PRO_NONE` and then go through the remaining if's? What is the relation between `PROT_NONE` and `prot = 0`?

Reply:
> What is the relation between `PROT_NONE` and `prot = 0`?

The short answer is that it depends on the OS. This is why I'd rather not make an assumption, even if on Linux `PROT_NONE` is `0`